dali/util/s3_filesystem.cc

```diff
@@ -0,0 +1,128 @@
+// Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/s3_filesystem.h"
+#include <aws/core/Aws.h>
+#include <aws/core/utils/stream/PreallocatedStreamBuf.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/HeadBucketRequest.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+#include <vector>
+#include "dali/core/format.h"
+#include "dali/core/nvtx.h"
+#include "dali/util/uri.h"
+
+namespace dali {
+
+namespace s3_filesystem {
+
+static const char kAllocationTag[] = "s3_filesystem";
+
+S3ObjectLocation parse_uri(const std::string& uri) {
+  auto parsed_uri = URI::Parse(uri);
+  if (parsed_uri.scheme() != "s3")
+    throw std::runtime_error("Not an S3 URI: " + uri);
+  S3ObjectLocation object_location;
+  object_location.bucket = parsed_uri.authority();
+  object_location.object = parsed_uri.path();
+  if (object_location.object.length() >= 1 && object_location.object[0] == '/')
+    object_location.object = object_location.object.substr(1);
+  return object_location;
+}
+
+S3ObjectStats get_stats(Aws::S3::S3Client* s3_client, const S3ObjectLocation& object_location) {
+  DomainTimeRange tr(make_string("get_stats @ ", object_location.object), DomainTimeRange::kOrange);
+  S3ObjectStats stats;
+  if (object_location.object.empty())
+    throw std::runtime_error("Object can't be empty");
+
+  Aws::S3::Model::HeadObjectRequest head_object_req;
+  head_object_req.SetBucket(object_location.bucket.c_str());
+  head_object_req.SetKey(object_location.object.c_str());
+  head_object_req.SetResponseStreamFactory(
+      []() { return Aws::New<Aws::StringStream>(kAllocationTag); });
+  auto head_object_outcome = s3_client->HeadObject(head_object_req);
+  stats.exists = head_object_outcome.IsSuccess();
```

Comment:
Does it means we can confuse error with file absence?

Reply:
the lack of existence is an error